packages/remix-react/fog-of-war.ts

```diff
@@ -0,0 +1,259 @@
+import type { Router } from "@remix-run/router";
+import { matchRoutes } from "@remix-run/router";
+import * as React from "react";
+import type { unstable_PatchRoutesOnMissFunction } from "react-router";
+
+import type { AssetsManifest, FutureConfig } from "./entry";
+import type { RouteModules } from "./routeModules";
+import { createClientRoutes } from "./routes";
+
+declare global {
+  interface Navigator {
+    connection?: { saveData: boolean };
+  }
+}
+
+type FogOfWarInfo = {
+  controller: AbortController | null;
+  // Currently rendered links that may need prefetching
+  nextPaths: Set<string>;
+  // Paths we know the client can already match, so no need to perform client-side
+  // matching or prefetching for them.  Just an optimization to avoid re-matching
+  // on a larger and larger route tree over time
+  knownGoodPaths: Set<string>;
+  // Routes the server was unable to match - don't ask for them again
+  known404Paths: Set<string>;
+};
+
+let fogOfWar: FogOfWarInfo | null = null;
+
+export function isFogOfWarEnabled(future: FutureConfig, isSpaMode: boolean) {
+  return future.unstable_fogOfWar === true && !isSpaMode;
+}
+
+export function initFogOfWar(
+  manifest: AssetsManifest,
+  routeModules: RouteModules,
+  future: FutureConfig,
+  isSpaMode: boolean,
+  basename: string | undefined
+): {
+  enabled: boolean;
+  patchRoutesOnMiss?: unstable_PatchRoutesOnMissFunction;
+} {
+  if (!isFogOfWarEnabled(future, isSpaMode)) {
+    return { enabled: false };
+  }
+
+  fogOfWar = {
+    controller: null,
+    nextPaths: new Set<string>(),
+    knownGoodPaths: new Set<string>(),
+    known404Paths: new Set<string>(),
+  };
+
+  return {
+    enabled: true,
+    patchRoutesOnMiss: async ({ path, patch }) => {
+      if (fogOfWar!.known404Paths.has(path)) return;
+      await fetchAndApplyManifestPatches(
+        [path],
+        fogOfWar!.known404Paths,
+        manifest,
+        routeModules,
+        future,
+        isSpaMode,
+        basename,
+        patch
+      );
+    },
+  };
+}
+
+export function useFogOFWarDiscovery(
+  router: Router,
+  manifest: AssetsManifest,
+  routeModules: RouteModules,
+  future: FutureConfig,
+  isSpaMode: boolean
+) {
+  React.useEffect(() => {
+    // Don't prefetch if not enabled or if the user has `saveData` enabled
+    if (
+      !isFogOfWarEnabled(future, isSpaMode) ||
+      navigator.connection?.saveData === true
+    ) {
+      return;
+    }
+
+    // Register a link href for patching
+    function registerPath(path: string | null) {
+      let { knownGoodPaths, known404Paths, nextPaths } = fogOfWar!;
+      if (path && !knownGoodPaths.has(path) && !known404Paths.has(path)) {
+        nextPaths.add(path);
+      }
+    }
+
+    // Fetch patches for all currently rendered links
+    async function fetchPatches() {
+      fogOfWar?.controller?.abort();
+
+      let lazyPaths = getFogOfWarPaths(fogOfWar!, router);
+      if (lazyPaths.length === 0) {
+        return;
+      }
```

Comment:
Short circuit there are no actively rendered paths to fetch - which means that they are all either matchable via the currently known routes, or they are a known 404 from a prior fetch.